shadowsocks/stream.go

```diff
@@ -21,11 +21,17 @@ import (
 	"fmt"
 	"io"
 	"sync"
+
+	"github.com/Jigsaw-Code/outline-ss-server/slicepool"
 )
 
 // payloadSizeMask is the maximum size of payload in bytes.
 const payloadSizeMask = 0x3FFF // 16*1024 - 1
 
+// Buffer pool used for encrypting and decrypting Shadowsocks streams.
+// The largest buffer we could need is for encrypting a max-length payload.
+var ssPool = slicepool.MakePool(payloadSizeMask + maxCipherOverhead())
```

Comment:
We could have a pool per cipher, but since they all have the same tag size, the simpler code doesn't waste any memory.

Reply:
Yes.  This approach is more efficient unless the differences in tag size become a large fraction of the total buffer size.